tonic/listen.go

```diff
@@ -0,0 +1,103 @@
+package tonic
+
+import (
+	"context"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+)
+
+var defaultOpts = []ListenOptFunc{
+	ListenAddr(":8080"),
+	CatchSignals(os.Interrupt, syscall.SIGTERM),
+	ShutdownTimeout(10 * time.Second),
+	ReadHeaderTimeout(5 * time.Second),
+	WriteTimeout(30 * time.Second),
+	KeepAliveTimeout(90 * time.Second),
+}
+
+func ListenAndServe(handler http.Handler, errorHandler func(error), opt ...ListenOptFunc) {
+
+	srv := &http.Server{Handler: handler}
+
+	listenOpt := &ListenOpt{Server: srv}
+
+	for _, o := range defaultOpts {
+		o(listenOpt)
+	}
+
+	for _, o := range opt {
+		o(listenOpt)
+	}
+
+	go func() {
+		err := srv.ListenAndServe()
+		if err != nil && err != http.ErrServerClosed && errorHandler != nil {
+			errorHandler(err)
```

Comment:
If `ListenAndServe` returns an error and all conditions are true, the `errorHandler` has the responsibility here to shutdown the application, otherwise it will get stuck at line 44, waiting for chan `quit` to be closed.

The chan `quit` could be declared before launching the goroutine that encloses `ListenAndServe`, and make sure that on error (different than `ErrServerClosed` than can only be the result of a call to `Shutdown()`), the channel is closed, so the execution can continue and the application shutdown properly.

Reply:
Good observations.

I agree with the problem statement, but not the proposed solution:

panic: send on closed channel

goroutine 5 [running]:
os/signal.process(0x4c4200, 0xc42008e000)
	/home/tschaffe/go/src/os/signal/signal.go:221 +0x184
os/signal.loop()

(sent a signal while the channel passed to Notify was already closed)

Reply:
yup, that's what I was doing.